File: src/waldiez/models/Agent/GroupManager/GroupManagerData.ts
/**
 * SPDX-License-Identifier: Apache-2.0
 * Copyright 2024 - 2025 Waldiez & contributors
 */
import {
    WaldiezAgentCodeExecutionConfig,
    WaldiezAgentData,
    WaldiezAgentHumanInputMode,
    WaldiezAgentLinkedTool,
    WaldiezAgentNestedChat,
    WaldiezAgentTerminationMessageCheck,
    WaldiezAgentUpdateSystemMessage,
} from "@waldiez/models/Agent/Common";
import { WaldiezAgentGroupManagerSpeakers } from "@waldiez/models/Agent/GroupManager/GroupSpeakers";
import { WaldiezTransitionTarget } from "@waldiez/models/common/Handoff";

/**
 * Waldiez Group Manager Agent Data.
 * @param humanInputMode - The human input mode of the agent ("NEVER" | "ALWAYS" | "TERMINATE")
 * @param systemMessage - The system message of the agent
 * @param codeExecutionConfig - The code execution configuration of the agent
 * @param agentDefaultAutoReply - The default auto reply of the agent
 * @param maxConsecutiveAutoReply - The maximum consecutive auto reply of the agent
 * @param termination - The termination message check of the agent
 * @param modelIds - The agent's model ids
 * @param tools - The tools available to the agent
 * @param parentId - The parent id of the agent
 * @param nestedChats - The nested chats of the agent
 * @param contextVariables - The context variables of the agent
 * @param updateAgentStateBeforeReply - The update agent state before reply of the agent
 * @param afterWork - The handoff transition after work of the agent
 * @param handoffs - The handoff / edge ids (used for ordering if needed)
 * @param groupName - The name of the group that the agent manages
 * @param maxRound - The maximum number of rounds for the group chat this agent manages
 * @param adminName - The admin name of the agent
 * @param speakers - The speakers of the agent
 * @param enableClearHistory - The enable clear history of the agent
 * @param sendIntroductions - The send introductions of the agent
 * @param initialAgentId - The id of the initial agent in the group
 * @see {@link WaldiezAgentData}
 * @see {@link WaldiezAgentLinkedTool}
 * @see {@link WaldiezAgentNestedChat}
 * @see {@link WaldiezAgentTerminationMessageCheck}
 * @see {@link WaldiezAgentGroupManagerSpeakers}
 * @see {@link WaldiezAgentHumanInputMode}
 * @see {@link WaldiezAgentCodeExecutionConfig}
 * @see {@link WaldiezAgentUpdateSystemMessage}
 * @see {@link WaldiezTransitionTarget}
 */
export class WaldiezAgentGroupManagerData extends WaldiezAgentData {
    maxRound: number;
    adminName: string | null;
    speakers: WaldiezAgentGroupManagerSpeakers;
    enableClearHistory?: boolean;
    sendIntroductions?: boolean;
    groupName?: string;
    initialAgentId?: string;

    constructor(
        props: {
            humanInputMode: WaldiezAgentHumanInputMode;
            systemMessage: string | null;
            codeExecutionConfig: WaldiezAgentCodeExecutionConfig;
            agentDefaultAutoReply: string | null;
            maxConsecutiveAutoReply: number | null;
            termination: WaldiezAgentTerminationMessageCheck;
            modelIds: string[];
            tools: WaldiezAgentLinkedTool[];
            parentId?: string | null;
            nestedChats: WaldiezAgentNestedChat[];
            contextVariables: Record<string, any>;
            updateAgentStateBeforeReply: WaldiezAgentUpdateSystemMessage[];
            afterWork: WaldiezTransitionTarget | null;
            handoffs: string[]; // handoff/edge ids
            maxRound: number;
            adminName: string | null;
            speakers: WaldiezAgentGroupManagerSpeakers;
            enableClearHistory?: boolean;
            sendIntroductions?: boolean;
            groupName?: string;
            initialAgentId?: string;
        } = {
            humanInputMode: "NEVER",
            systemMessage: null,
            codeExecutionConfig: false,
            agentDefaultAutoReply: null,
            maxConsecutiveAutoReply: null,
            termination: {
                type: "none",
                keywords: [],
                criterion: null,
                methodContent: null,
            },
            modelIds: [],
            tools: [],
            parentId: undefined,
            nestedChats: [
                {
                    messages: [],
                    triggeredBy: [],
                    condition: {
                        conditionType: "string_llm",
                        prompt: "",
                    },
                    available: {
                        type: "none",
                        value: "",
                    },
                },
            ],
            contextVariables: {},
            updateAgentStateBeforeReply: [],
            afterWork: null,
            handoffs: [],
            maxRound: 20,
            adminName: null,
            speakers: {
                selectionMethod: "auto",
                selectionCustomMethod: "",
                maxRetriesForSelecting: null,
                selectionMode: "repeat",
                allowRepeat: true,
                allowedOrDisallowedTransitions: {},
                transitionsType: "allowed",
                order: [],
            },
            enableClearHistory: undefined,
            sendIntroductions: undefined,
            groupName: undefined,
            initialAgentId: undefined,
        },
    ) {
        props.parentId = undefined;
        super(props);
        this.maxRound = props.maxRound;
        this.adminName = props.adminName;
        this.speakers = props.speakers;
        this.enableClearHistory = props.enableClearHistory;
        this.sendIntroductions = props.sendIntroductions;
        this.groupName = props.groupName;
        this.initialAgentId = props.initialAgentId;
    }
}

File: waldiez/models/agents/group_manager/group_manager_data.py
# SPDX-License-Identifier: Apache-2.0.
# Copyright (c) 2024 - 2025 Waldiez and contributors.
"""Group chat manager data."""

from typing import Optional

from pydantic import Field
from typing_extensions import Annotated, Literal

from ..agent import WaldiezAgentData
from .speakers import WaldiezGroupManagerSpeakers


class WaldiezGroupManagerData(WaldiezAgentData):
    """Group chat manager data class.

    The data for an agent with `human_input_mode` set to "NEVER" as default.
    and the chat group's related extra properties.
    See the parent's docs (`WaldiezAgentData`) for the rest of the properties.

    Attributes
    ----------
    human_input_mode : Literal["ALWAYS", "NEVER", "TERMINATE"]
        The human input mode, Defaults to `NEVER`
    max_round : int
        The maximum number of rounds to have in the group. Defaults to 20.
    admin_name : Optional[str]
        The name of the group's admin.
        Make sure you use a name of an agent in the group.
    speakers : WaldiezGroupManagerSpeakers
        The rules for the speaker selection and repetition
    enable_clear_history : Optional[bool]
        Enable clearing the history in the chat group.
    send_introductions : bool
        Send the group members' introductions.
    group_name : Optional[str]
        The name of the group.
    initial_agent_id: str
        The ID of the initial agent to be used in the group.
    """

    human_input_mode: Annotated[
        Literal["ALWAYS", "NEVER", "TERMINATE"],
        Field(
            default="NEVER",
            title="Human input mode",
            description="The human input mode, Defaults to `NEVER`",
            alias="humanInputMode",
        ),
    ]
    max_round: Annotated[
        int,
        Field(
            default=20,
            title="Max round",
            description="The maximum number of rounds to have in the group.",
            alias="maxRound",
        ),
    ]
    admin_name: Annotated[
        Optional[str],
        Field(
            default=None,
            title="Group Admin name",
            description=(
                "The name of the group's admin. "
                "Make sure you use a name of an agent in the group."
            ),
            alias="adminName",
        ),
    ]
    speakers: Annotated[
        WaldiezGroupManagerSpeakers,
        Field(
            title="Speakers",
            description="The rules for the speaker selection and repetition",
            default_factory=WaldiezGroupManagerSpeakers,  # pyright: ignore
        ),
    ]
    enable_clear_history: Annotated[
        Optional[bool],
        Field(
            default=None,
            title="Enable clear history",
            description="Enable clearing the history in the chat group.",
            alias="enableClearHistory",
        ),
    ]
    send_introductions: Annotated[
        bool,
        Field(
            default=False,
            title="Send Introductions",
            description="Send the group members' introductions.",
            alias="sendIntroductions",
        ),
    ]
    group_name: Annotated[
        Optional[str],
        Field(
            default=None,
            title="Group name",
            description="The name of the group.",
            alias="groupName",
        ),
    ]
    initial_agent_id: Annotated[
        str,
        Field(
            ...,
            title="Initial agent ID",
            description="The ID of the agent that starts the conversation.",
            alias="initialAgentId",
        ),
    ]
